app/src/main/java/androidx/viewpager2/integration/NestedScrollableHost.java

```diff
@@ -0,0 +1,133 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Source: https://github.com/android/views-widgets-samples/blob/87e58d1/ViewPager2/app/src/main/java/androidx/viewpager2/integration/testapp/NestedScrollableHost.kt
+ */
+
+package androidx.viewpager2.integration;
+
+import android.content.Context;
+import android.util.AttributeSet;
+import android.view.MotionEvent;
+import android.view.View;
+import android.view.ViewConfiguration;
+import android.widget.FrameLayout;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.viewpager2.widget.ViewPager2;
+
+import static androidx.viewpager2.widget.ViewPager2.ORIENTATION_HORIZONTAL;
+
+/**
+ * Layout to wrap a scrollable component inside a ViewPager2. Provided as a solution to the problem
+ * where pages of ViewPager2 have nested scrollable elements that scroll in the same direction as
+ * ViewPager2. The scrollable element needs to be the immediate and only child of this host layout.
+ *
+ * <p>This solution has limitations when using multiple levels of nested scrollable elements
+ * (e.g. a horizontal RecyclerView in a vertical RecyclerView in a horizontal ViewPager2).</p>
+ */
+class NestedScrollableHost extends FrameLayout {
+
+    public NestedScrollableHost(@NonNull Context context) {
+        super(context);
+        touchSlop = ViewConfiguration.get(context).getScaledTouchSlop();
+    }
+
+    public NestedScrollableHost(@NonNull Context context, @Nullable AttributeSet attrs) {
+        super(context, attrs);
+        touchSlop = ViewConfiguration.get(context).getScaledTouchSlop();
+    }
+
+    private int touchSlop = 0;
+    private float initialX = 0f;
+    private float initialY = 0f;
+
+    private ViewPager2 getParentViewPager() {
+        View v = (View) getParent();
+        while (v != null && !(v instanceof ViewPager2)) {
+            v = (View) v.getParent();
+        }
+        return v == null ? null : (ViewPager2) v;
+    }
+
+    private View getChild() {
+        return getChildCount() > 0 ? getChildAt(0) : null;
+    }
+
+    private boolean canChildScroll(int orientation, float delta) {
+        int direction = (int) -Math.copySign(1, delta);
+        View child = getChild();
+        if (child == null) {
+            return false;
+        }
+        switch (orientation) {
+            case 0:
+                return child.canScrollHorizontally(direction);
+            case 1:
+                return child.canScrollVertically(direction);
+            default:
+                return false;
+        }
+    }
+
+    public boolean onInterceptTouchEvent(MotionEvent e) {
+        handleInterceptTouchEvent(e);
+        return super.onInterceptTouchEvent(e);
+    }
+
+    private void handleInterceptTouchEvent(MotionEvent e) {
```

Comment:
Scrolling generally works fine now but it sometimes feels a bit strange.

- It is really easy to accidentally switch to another page when trying to scroll up/down. I think the sensitivity should be inverted: It should be easy to scroll up/down and harder to switch the page. Not sure how that would work, I have not looked at the code too closely yet.
- Sometimes, flinging (move finger fast, then lift and the view continues to scroll) does not work. This might be caused by the thing above.

Reply:
stuck with it, guess need your help :)

Reply:
wrapping it in NestedScrollView make it not scroll easily but causes more harm elsewhere, make it similar to bottomsheet shownote issue, I think ideally both webviews should be wrap content to make nestedscrollview handle their scrolling but that seems also tricky.

Reply:
> I think ideally both webviews should be wrap content to make nestedscrollview handle their scrolling

This is not ideal. Wrapping in `nestedscrollview` causes me headache already because it increases the memory footprint of the WebView. Setting it to `wrap_content` makes it render the whole page, even the parts that are not visible. If the page contains many images, this can lead to crashes. So ideally, I want to go completely without the `nestedscrollview` hack.

> after playing some hours with it, no progress, guess need your help :)

Maybe implementing `NestedScrollingChild3` could be a starting point. I do not know much about that topic either, so I can not help too much, unfortunately.